File: services/src/main/java/com/app/appointment_booking_system/service/AppointmentService.java
package com.app.appointment_booking_system.service;

import com.app.appointment_booking_system.dto.CreateAppointmentRequest;
import com.app.appointment_booking_system.dto.UpdateAppointmentRequest;
import com.app.appointment_booking_system.model.Appointment;
import com.app.appointment_booking_system.model.Appointment.AppointmentStatus;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

public interface AppointmentService {

    // CREATE
    Appointment createAppointment(CreateAppointmentRequest request);

    // READ
    List<Appointment> getAllAppointments();
    
    Optional<Appointment> getAppointmentById(Long id);
    
    List<Appointment> getAppointmentsByUserId(Long userId);
    
    List<Appointment> getAppointmentsByStatus(AppointmentStatus status);
    
    List<Appointment> getAppointmentsByDate(LocalDate date);
    
    List<Appointment> getAppointmentsInRange(LocalDateTime startDate, LocalDateTime endDate);
    
    // Check availability for a specific time slot
    boolean isTimeSlotAvailable(LocalDateTime dateTime);

    // UPDATE
    Appointment updateAppointment(Long id, UpdateAppointmentRequest request);

    // CANCEL
    Appointment cancelAppointment(Long id);

    // COMPLETE
    Appointment completeAppointment(Long id);
}

File: appointment-ui/src/api/appointmentService.ts
import api from "./api";

export enum AppointmentStatus {
  SCHEDULED = "SCHEDULED",
  CANCELLED = "CANCELLED",
  COMPLETED = "COMPLETED",
}

export interface Appointment {
  id?: number;
  user: {
    id: number;
    name: string;
    phoneNumber: string;
    email: string;
  };
  appointmentDateTime: string;
  notes?: string;
  status: AppointmentStatus;
  createdAt?: string;
  updatedAt?: string;
}

export interface CreateAppointmentRequest {
  userId: number;
  appointmentDateTime: string;
  notes?: string;
}

export interface UpdateAppointmentRequest {
  userId?: number;
  appointmentDateTime?: string;
  notes?: string;
  status?: AppointmentStatus;
}

// GET all appointments
export const getAllAppointments = async () => {
  const response = await api.get<Appointment[]>("/appointments");
  return response.data;
};

// GET appointment by ID
export const getAppointmentById = async (id: number) => {
  const response = await api.get<Appointment>(`/appointments/${id}`);
  return response.data;
};

// GET appointments by user ID
export const getAppointmentsByUserId = async (userId: number) => {
  const response = await api.get<Appointment[]>(`/appointments/user/${userId}`);
  return response.data;
};

// GET appointments by status
export const getAppointmentsByStatus = async (status: AppointmentStatus) => {
  const response = await api.get<Appointment[]>(`/appointments/status/${status}`);
  return response.data;
};

// GET appointments by date
export const getAppointmentsByDate = async (date: string) => {
  const response = await api.get<Appointment[]>(`/appointments/date/${date}`);
  return response.data;
};

// GET appointments in date range
export const getAppointmentsInRange = async (startDate: string, endDate: string) => {
  const response = await api.get<Appointment[]>(`/appointments/range`, {
    params: { startDate, endDate },
  });
  return response.data;
};

// CHECK availability
export const checkAvailability = async (dateTime: string) => {
  const response = await api.get<boolean>(`/appointments/availability`, {
    params: { dateTime },
  });
  return response.data;
};

// CREATE appointment
export const createAppointment = async (request: CreateAppointmentRequest) => {
  const response = await api.post<Appointment>("/appointments", request);
  return response.data;
};

// UPDATE appointment
export const updateAppointment = async (id: number, request: UpdateAppointmentRequest) => {
  const response = await api.put<Appointment>(`/appointments/${id}`, request);
  return response.data;
};

// CANCEL appointment
export const cancelAppointment = async (id: number) => {
  const response = await api.put<Appointment>(`/appointments/${id}/cancel`);
  return response.data;
};

// COMPLETE appointment
export const completeAppointment = async (id: number) => {
  const response = await api.put<Appointment>(`/appointments/${id}/complete`);
  return response.data;
};

// DELETE appointment
export const deleteAppointment = async (id: number) => {
  await api.delete(`/appointments/${id}`);
};
